### ISSUE TITLE
FRAC() does not work correctly with big numbers like 4.5678879E8

### ISSUE THREAD

**Issue by olimaticer** [RFO-BASIC/Basic#234]
_**Solution:**_
	**private boolean executeMF_FRAC() {**					// 2014-03-16 gt
		**if (!evalNumericExpression()) return false;**
		//Old version does not work correctly with big numbers like 4.5678879E8
		//String str = EvalNumericExpressionValue.toString();
		//String sgn = (EvalNumericExpressionValue < 0) ? "-" : "";
		//int point = str.indexOf('.');
		//EvalNumericExpressionValue = (point < 0) ? 0.0 : Double.parseDouble(sgn + str.substring(point));
		//vv2017-04-12gt
		**EvalNumericExpressionValue = EvalNumericExpressionValue - EvalNumericExpressionValue.longValue();**
		//^^2017-04-12gt
		**return true;**
	**}**

**Comment by jMarcS** [comment]
Nice catch - that's a very ugly bug.

Your test case, 4.5678879E8, is bad -- it is an integer. Does your new FRAC() work with numbers that have a non-zero fractional part?

Do you know why the old code fails?

Subtracting a large long int from a double precision float is likely to have bad rounding errors. The old code does not do any arithmetic (except what Java does in converting a string). It is very good at giving you back your actual decimal digits. Is your new code as good?

Have you considered a better fix, one that has the benefits of the old code but without the bug?

Where does your new code stop working? What should we put in the manual about the accuracy and precision of FRAC()?

**Comment by olimaticer** [comment]
Hi Marc,
I stumbled over this:
t = time()
print t
print t/1000
print frac(t/1000)
? t/1000 - int(t/1000) %Workaround

It fails, because with 4.5678879E8 Frac() returns 0.5678879
The issue is the result of .toString().
The solution works with positive and negative numbers.
 Result and precision are the same as the workaround.
/ Gregor

**Comment by olimaticer** [comment]
Hi Marc,

You are asking the right questions.
We can make it good, but we could make it better.


private boolean executeMF_FRAC() {	// 2014-03-16 gt
if (!evalNumericExpression()) return false;
String str = EvalNumericExpressionValue.toString();
If(str.indexOf("E") > -1){
EvalNumericExpressionValue = EvalNumericExpressionValue - EvalNumericExpressionValue.longValue();
}else{
String sgn = (EvalNumericExpressionValue < 0) ? "-" : "";
int point = str.indexOf('.');
EvalNumericExpressionValue = (point < 0) ? 0.0 : Double.parseDouble(sgn + str.substring(point));
}
return true;
}
/Gregor

**Comment by jMarcS** [comment]
We cannot use code based on subtraction. Here are some test cases:
`frac(1.1)` returns `0.10000000000000009`
`frac(1.1234` returns `0.12339999999999995`
`frac(4.56788791234E8)` returns `0.23400002717971802`

>> Where does your new code stop working?

> Result and precision are the same as the workaround.

I don't understand what that means.

Here are some reference information I found:
> double: 64 bits (8 bytes) where 52 bits are used for the mantissa (15 to 17 decimal digits, about 16 on average). 11 bits are used for the exponent and 1 bit is the sign bit.

> long -9,223,372,036,854,775,808 to 9,223,372,036,854,775,807
float -3.402823e38 to 3.402823e38
double -1.79769313486232e308 to 1.79769313486232e308

Double precision floating point can't hold as many digits as long integers. Have you tested the boundaries? I need to know where the algorithm stops working so I can put it in the manual.

I considered using BigDecimal numbers, but that only works if the frac() argument is a literal. The argument can be any numerical expression, so we are stuck with parsing a double.

We could fix the rounding errors, but it is better to use a Java API that already does the rounding. One possibility is `String.format()`. In the original code, replace
````String str = EvalNumericExpressionValue.toString();````
with
````String str = String.format(Locale.US,"%.16f", EvalNumericExpressionValue);````

I don't know exactly what the '%f" conversion values should be. "%.16f" seems to work, but I have not tested it well. I chose 16 as a starting point. Perhaps 18 is best, because it includes all possible doubles. Perhaps 15 is best, because it excludes digits that are uncertain. We need to find some test cases that show the difference, then decide what to do.

**Comment by jMarcS** [comment]
A quick test shows there is an internal limit in `String.format`:
`String.format(Locale.US,"%.18f", 0.12345678901234567)` is `0.123456789012346000`
The result is correctly rounded to 15 digits. Java has made the decision for us.

That's not perfect, as shown in this BASIC! program:
`a = 0.12345678901234567`
`PRINT a`
`PRINT FRAC(a)`
The result is:
`0.12345678901234566`
`0.123456789012346`

The problem is that `String.format` and `Double.toString` are not consistent. Worse, `Double.toString` does not create the correct string. The 17th digit is wrong.

**Comment by olimaticer** [comment]
String s = String.valueOf(EvalNumericExpressionValue);
s returns 0.12345678901234566

With my last code suggestion I got these results:

a = 0.12345678901234567
PRINT a
PRINT FRAC(a)
PRINT a - INT(a)
k = 10^21
PRINT (a*k - INT(a)*k) / k
The result is:
0.12345678901234566
0.12345678901234566
0.12345678901234566
0.12345678901234566

a = 1.4
PRINT a
PRINT FRAC(a)
PRINT a - INT(a)
k = 10^21
PRINT (a*k - INT(a)*k) / k
The result is:
1.4
0.4
0.39999999999999999
0.4

a = time()/1000
PRINT a
PRINT FRAC(a)
PRINT a - INT(a)
k = 10^21
PRINT (a*k - INT(a)*k) / k
The result is:
1.492312809399
0.3989999294281006
0.3989999294281006
0,39900006816158634

**Comment by jMarcS** [comment]
First case: Consistent answers. Consistently wrong, but consistent.
Second case: Works correctly because it does not use the subtraction method in your code.
Third case: Unusable. The correct result is 0.399. 

The third case puzzles me. It is accurate only to 7 decimal places. It should be 15. There must be a single-precision float somewhere in the calculation. This messes up any attempt to fix the subtraction result by rounding.

For comparison, frac() based on String.Format gets these results:

0.12345678901234567  a
0.123456789012346      frac(a): correct, but only 15 digits

1.4                                  a
0.4                                  frac(a): correct

1.492352936901E9       a: current time
0.901                             frac(a): correct

I intend to write in the manual that frac() uses no more than 15 significant digits. When working with Java Double, nothing more than that is reliable.

I wonder if we should fix PRINT, too? Up to now we've left the problem of error margin in floating point comparison to the user, but I don't think we've told the users! Some other versions of BASIC will always round, so `0.39999999999999999` prints as `0.4`. They also print `1.0` as `1`. Even better, some allow a runtime option to set the output precision.

**Comment by olimaticer** [comment]
Yesterday I tried also **Math.rint**(EvalNumericExpressionValue); (Round to Integer) in the INT() function
with PRINT a - INT(a). I got an absolutly wrong result.

> 0.12345678901234567 a
> 0.123456789012346 frac(a): correct, but only 15 digits

A correct result, but the variance is higher.
Try also with direct input:
0.12345678901234560
0.12345678901234561
0.12345678901234562
0.12345678901234563
0.12345678901234564
0.12345678901234565
0.12345678901234566
0.12345678901234567
0.12345678901234568
0.12345678901234569
and also:
a = 0.1234567890123456
k = 10^21
FOT i = 0 TO 9
   a = a + i * 10^-17
   PRINT a
   PRINT FRAC(a)
   PRINT a - INT(a)
   PRINT (a*k - INT(a)*k) / k
NEXT i

I think we discuss on a high level, but a addition or a subtraction returns often an incorrect result.
And some math libs are in some cases also unusable.
Should we not prefer a solution with a variance so less as possible?
Only in the case of an E in the number we should use frac() based on String.Format.
But we have also to explain the difference between FRAC(a) and  a - INT(a) in the manual.

When I have invented the ten phases semiconductor, I will inform you:-)

/ Gregor

**Comment by olimaticer** [comment]
> I wonder if we should fix PRINT, too? Up to now we've left the problem of error margin in floating point comparison to the user, but I don't think we've told the users!
Maybe we should talk about the exponential shifting hack in the range of the bounds of Double numbers, too.
k = 10^21
c =  (a*k - b*k) / k

**Comment by olimaticer** [comment]
For SQR(2)*10^12
EvalNumericExpressionValue.toString
returns
**1.41213562373095**2E12

String.format(Locale.US ,"%.18f", EvalNumericExpressionValue);
returns 
**141213562373**.10000000000000000

new BigDecimal(EvalNumericExpressionValue).toPlainString();
returns
**1141213562373.095**2184375

correct is
**1141213562373.095048801688724210...**

So the last one has the lowest variance.
We should use BigDecimal if an "E" is given.

Gregor

**Comment by jMarcS** [comment]
No. If we use `BigDecimal`, then `frac(sqr(2)*10^12)` returns `0.0952184375`. That's pure nonsense. Anything beyond the 15th digit of the argument is suspect and should not be printed.

The correct value of `sqr(2)` to 17 digits is `1.4142135623730950`. Rounded to 15 digits that is `1.41421356237310`. The correct return value of `frac(sqr(2)*10^12)` is `0.10`, which BASIC! prints as `0.1`.

Yes, I realize that `EvalNumericExpressionValue.toString` actually returns 16 correct digits for this particular number, so printing `0.095` would be accurate. Most double precision numbers are correct in the 16th digit. But there are numbers for which only 15 digits are correct, so that's got to be our limit.
